### PR TITLE
How to Choose Levels of Logging

### LINKED ISSUE #1438: How to Choose Levels of Logging
**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

### Brief Summary: 
Describe the what, why, and how of your content idea in 2-5 sentences.

Application logging is one of the most important things you can do in your code when facilitating production support. Your log files serve as a sort of archaeological record of what your codebase did in production on earth. Each entry in a log file has essential information, including a timestamp, contextual information, and a message. And most importantly, something called a logging level. This article describes different logging levels.  

### Key Takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?

1. What is a logging level?
2. Background to logging levels 
3. The most common logging levels (DEBUG, ERROR, FATAL, INFO, etc.) and how to choose them. 

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A.')
https://www.tutorialspoint.com/log4j/log4j_logging_levels.htm

### PR DESCRIPTION
**Pull request by EKahuha**
this fixes #1438

### Article Checklist
**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process, quality, and style-once published, you can then submit any number of articles.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

File: articles/how-to-choose-levels-of-logging/index.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+With DEBUG, you are giving diagnostic information in a detailed manner. It is verbose and has more information than you would need when using the application. DEBUG logging level is used to fetch information needed to diagnose, troubleshoot, or test an application. This ensures the smooth running of an application.
+
+#### TRACE
+
+TRACE log level captures all details about the behavior of the application. It is mostly diagnostic and is more granular and finer than DEBUG log level. This log level is used in situations where one needs to see what happens in your application or what happens in the third-party library used. You can use the TRACE log level to query parameters in the code or interpret the algorithm&#39;s steps.
+
+#### ALL
+
+This log level logs any logging levels that are defined. It logs everything and includes custom logging levels as well. It is merely the unification of all other logging levels where the whole lot of loggings will be logged.
+
```
Comment (geoffrey45) [inline articles/how-to-choose-levels-of-logging/index.md:50]: Please rewrite this last sentence using simpler and commonly used terms. Ie. In "merely the unification" and "whole lot of loggings will be logged".

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
@EKahuha Great article. Just some tiny issues. Please make the changes attached below.

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/how-to-choose-levels-of-logging/index.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+### Background of logging levels
+
+Logging levels distinguish various log events from each other. They are a way of filtering important information about your system's state and reduce information noise or alert fatigue.
+
+Log levels were introduced in the 1980s in a project named [Sendmail](https://en.wikipedia.org/wiki/Sendmail). The project owners needed a logging solution that eventually led to the creation of a Syslog server and the idea of severity levels.
+
+Eric Allman developed the Sendmail project that eventually needed a logging solution. This is where the System Logging Protocol ([Syslog](https://www.paessler.com/it-explained/syslog#)) was introduced. Ultimately, Syslog was embraced by numerous applications and has become a standard protocol for sending system log and event messages to the Syslog server.
```
Comment (geoffrey45) [inline articles/how-to-choose-levels-of-logging/index.md:15]: You have some repetition here. The line: `Eric Allman developed the Sendmail project that eventually needed a logging solution.` repeats line 13.

Also, provide the Syslog link and full form of the word in the first appearance of the word.

File: articles/how-to-choose-levels-of-logging/index.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+#### OFF
+
+This log level does not log anything. This OFF level is used for turning off logging and is the greatest possible rank. With this log level, nothing gets logged at all.
+
+### Do log levels work on their own?
+
+Logging levels do not work well on their own. They mostly serve information purposes in the log message. They provide information on how crucial an event is. You can limit the information written to the log level by combining the log level with a logging framework.
+
+Setting a specific log level in the logging framework ensures less important logging levels are ignored. For example, if the logging framework has the root log level as WARN, you only get events with FATAL level, ERROR level, and WARN level.
+
```
Comment (geoffrey45) [inline articles/how-to-choose-levels-of-logging/index.md:62]: Looks like the log levels are ordered. Consider turning them into an ordered list. Also, state how they are ordered. ie. according to what criteria and in what order.

**Review by geoffrey45 — COMMENTED** [review]
@EKahuha Almost done with the review. Just some minor suggestions. Please look at the comments below.

**Comment by hectorkambow** [comment]
@geoffrey45  
just wanted to get a status update on this article

**Comment by hectorkambow** [comment]
> @hectorkambow I'm checking the article for the last time before I submit it for final review. I had suggested some minor changes from the author which came through almost 2 hours ago.

Understood - thank you. 
Wasn't sure when the author changes came in - no worries or rush. 
Thank you v much 🚀

**Comment by hectorkambow** [comment]
@EKahuha  Great work Gentlemen 

Very minor edits on my end 👍 

cc @geoffrey45  🚀

**Pull request merged**